plugins/circuitbreaker/main.go

```diff
@@ -0,0 +1,524 @@
+	FailureRateThreshold                  float64           // Failure rate threshold (0.0 to 1.0)
+	SlowCallRateThreshold                 float64           // Slow call rate threshold (0.0 to 1.0)
+	SlowCallDurationThreshold             time.Duration     // Duration threshold for slow calls
+	MinimumNumberOfCalls                  int               // Minimum calls before evaluation
+	SlidingWindowType                     SlidingWindowType // Type of sliding window
+	SlidingWindowSize                     int               // Size of sliding window
+	PermittedNumberOfCallsInHalfOpenState int               // Calls allowed in half-open state
+	MaxWaitDurationInHalfOpenState        time.Duration     // Wait time before half-open transition
+}
+
+// CircuitBreaker implements the Bifrost plugin interface to provide circuit breaker
+// functionality. It maintains separate circuit states for each provider and uses
+// sliding windows to track call metrics.
+type CircuitBreaker struct {
+	config CircuitBreakerConfig
+
+	// Per-provider circuit states
+	mu        sync.RWMutex
+	providers map[schemas.ModelProvider]*ProviderCircuitState
+}
+
+// CircuitBreakerMetrics provides observability data for a circuit breaker instance.
+type CircuitBreakerMetrics struct {
+	State                  CircuitState // Current circuit state
+	FailureRate            float64      // Current failure rate
+	SlowCallRate           float64      // Current slow call rate
+	TotalCalls             int          // Total calls in window
+	FailedCalls            int          // Failed calls in window
+	SlowCalls              int          // Slow calls in window
+	StateTransitionTime    time.Time    // Time of last state transition
+	InFlightCalls          int          // Currently in-flight calls
+	HalfOpenCallsAttempted int          // Calls attempted in half-open state
+	HalfOpenCallsPermitted int          // Calls permitted in half-open state
+	HalfOpenSuccesses      int          // Successful calls in half-open state
+}
+
+// Context keys for storing circuit breaker data in request context.
+type contextKey string
+
+const (
+	callStartTimeKey contextKey = "circuitbreaker_call_start_time"
+	circuitStateKey  contextKey = "circuitbreaker_circuit_state"
+	providerKey      contextKey = "circuitbreaker_provider"
+)
+
+// NewDefaultCircuitBreakerPlugin creates a circuit breaker plugin with default configuration
+func NewDefaultCircuitBreakerPlugin() *CircuitBreaker {
+	cb, err := NewCircuitBreakerPlugin(DefaultConfig())
+	if err != nil {
+		// This should never happen with default config, but if it does, panic
+		panic(fmt.Sprintf("failed to create circuit breaker with default config: %v", err))
+	}
+	return cb
+}
+
+
+// NewCircuitBreakerPlugin creates a new circuit breaker plugin with the given configuration.
+// It validates the configuration and uses default values for any invalid parameters.
+func NewCircuitBreakerPlugin(config CircuitBreakerConfig) (*CircuitBreaker, error) {
```

Comment:
have only one NewCircuitBreakerPlugin method, and add default config if its not present